ASTNode.h

```diff
@@ -1,735 +1,794 @@
+    };
+
+    ASTNode(int type = NODE_INVALID) : m_refs(), m_type(type), m_processed() { }
+    virtual ~ASTNode() { }
+
+    int type() const { return internalGetType(this); }
+
+    bool processed() const { return m_processed; }
+    void setProcessed() { m_processed = true; }
+
+private:
+    int m_refs;
+    int m_type;
+    bool m_processed;
+
+    // Hack to make clang happy :(
+    static int internalGetType(const ASTNode *node)
+    {
+        return node ? node->m_type : NODE_INVALID;
+    }
+
+    static void internalAddRef(ASTNode *node)
+    {
+        if (node)
+            ++node->m_refs;
+    }
+
+    static void internalDelRef(ASTNode *node)
+    {
+        if (node && --node->m_refs == 0)
+            delete node;
+    }
+
+public:
+    void addRef() { internalAddRef(this); }
+    void delRef() { internalDelRef(this); }
+};
+
+
+class ASTNodeList : public ASTNode {
+public:
+    typedef std::list<PycRef<ASTNode>> list_t;
+
+    ASTNodeList(list_t nodes)
+        : ASTNode(NODE_NODELIST), m_nodes(std::move(nodes)) { }
+
+    const list_t& nodes() const { return m_nodes; }
+    void removeFirst();
+    void removeLast();
+    void append(PycRef<ASTNode> node) { m_nodes.emplace_back(std::move(node)); }
+
+protected:
+    ASTNodeList(list_t nodes, ASTNode::Type type)
+        : ASTNode(type), m_nodes(std::move(nodes)) { }
+
+private:
+    list_t m_nodes;
+};
+
+
+class ASTChainStore : public ASTNodeList {
+public:
+    ASTChainStore(list_t nodes, PycRef<ASTNode> src)
+        : ASTNodeList(nodes, NODE_CHAINSTORE), m_src(std::move(src)) { }
+    
+    PycRef<ASTNode> src() const { return m_src; }
+
+private:
+    PycRef<ASTNode> m_src;
+};
+
+
+class ASTObject : public ASTNode {
+public:
+    ASTObject(PycRef<PycObject> obj)
+        : ASTNode(NODE_OBJECT), m_obj(std::move(obj)) { }
+
+    PycRef<PycObject> object() const { return m_obj; }
+
+private:
+    PycRef<PycObject> m_obj;
+};
+
+
+class ASTUnary : public ASTNode {
+public:
+    enum UnOp {
+        UN_POSITIVE, UN_NEGATIVE, UN_INVERT, UN_NOT
+    };
+
+    ASTUnary(PycRef<ASTNode> operand, int op)
+        : ASTNode(NODE_UNARY), m_op(op), m_operand(std::move(operand)) { }
+
+    PycRef<ASTNode> operand() const { return m_operand; }
+    int op() const { return m_op; }
+    virtual const char* op_str() const;
+
+protected:
+    int m_op;
+
+private:
+    PycRef<ASTNode> m_operand;
+};
+
+
+class ASTBinary : public ASTNode {
+public:
+    enum BinOp {
+        BIN_ATTR, BIN_POWER, BIN_MULTIPLY, BIN_DIVIDE, BIN_FLOOR, BIN_MODULO,
+        BIN_ADD, BIN_SUBTRACT, BIN_LSHIFT, BIN_RSHIFT, BIN_AND, BIN_XOR,
+        BIN_OR, BIN_LOG_AND, BIN_LOG_OR, BIN_MAT_MULTIPLY,
+        /* Inplace operations */
+        BIN_IP_ADD, BIN_IP_SUBTRACT, BIN_IP_MULTIPLY, BIN_IP_DIVIDE,
+        BIN_IP_MODULO, BIN_IP_POWER, BIN_IP_LSHIFT, BIN_IP_RSHIFT, BIN_IP_AND,
+        BIN_IP_XOR, BIN_IP_OR, BIN_IP_FLOOR, BIN_IP_MAT_MULTIPLY,
+    };
+
+    static BinOp getBinOpFromOperand(int operand) {

```

Comment:
This function is long enough it's probably worth moving the implementation to ASTNode.cpp rather than including it inline.

Reply:
True it was mostly for testing I will update this